src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/SequencePipeReader.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class SequencePipeReader : PipeReader
+    {
+        private ReadOnlySequence<byte> _sequence;
+        private bool _isReaderCompleted;
+
+        private CancellationTokenSource? _internalTokenSource;
+        private readonly object _lock = new object();
+
+        public SequencePipeReader(ReadOnlySequence<byte> sequence)
+        {
+            _sequence = sequence;
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        /// <inheritdoc />
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            ThrowIfCompleted();
+
+            // Fast path: did we consume everything?
+            if (consumed.Equals(_sequence.End))
+            {
+                _sequence = ReadOnlySequence<byte>.Empty;
+                return;
+            }
+
+            _sequence = _sequence.Slice(consumed);
+        }
+
+        private CancellationTokenSource InternalTokenSource
```

Comment:
This isn't needed.

Reply:
Per the documentation, you need to return cancelled on *next* request if the cancel has been called. I initially didn't have this but added it for that reason.

Reply:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs#L74

Reply:
I think it's fine for `CancelPendingRead` to noop if ReadAsync is always completed.

Reply:
That's a contract change according to the docs though? This is explicitly tested in the stream variant with TryRead. I didn't want to assume I could do that.

Reply:
While CancelPendingRead does feel a bit useless given ReadAsync is always completed, not returning a canceled ReadResult from subsequent calls to ReadAsync/TryRead breaks the contract. I don't think we need a CancellationTokenSource. Using Interlocked.Exchange similar to how we do in Kestrel's [Http1UpgradeMessageBody](https://github.com/dotnet/aspnetcore/blob/main/src/Servers/Kestrel/Core/src/Internal/Http/Http1UpgradeMessageBody.cs#L47) should be fine.

Reply:
I give you permission.

Reply:
In fact! I would update the docs to say that it may not fire if the ReadResult is already completed.

Reply:
@halter73 and I had some discussion offline. Here's how I feel about this. I think supporting cancellation is pointless but I can be convinced with a reasonable scenario. Also the current implementation is overkill as ReadAsync never waits. I'd love to understand how code could react to this, maybe I'm missing a simple test case

Reply:
Personally I don't have a dog in this fight (I believe that is the expression 😛). I added it because the API docs explicitly described this behaviour and there were unit tests that tested this behaviour, and I would consider not doing it a breaking change.

That being the case, having that breaking change (and announcing it I guess?) is perfectly fine for me.

@halter73 changing to atomics should be pretty straight forward. Not doing so was one of the reasons I created this PR in a draft state. If it's decided that the cancellation behaviour should remain as described in the docs, I will update the code to do that.

Reply:
I'll wait for the simplified version of cancellation

Reply:
I replaced it with some `Interlocked.Exchange` instructions and a single integer to keep track of whether or not cancellation has been requested. Let me know what you think :).